### ISSUE TITLE
Gradient computation issues in the `caustics`

### ISSUE THREAD

**Issue by ShotaMiyazaki94** [ShotaMiyazaki94/microjax#11]
I integrated the newest version of `caustics` into microJAX as-is for a CPU contour-integration path. Magnification curves compute successfully, but the gradient calculations are numerically unstable, and the runtime doesn’t appear fully optimized in this setup.
If we upgrade this approach as a CPU implementation within microJAX, we’ll need to address these points for stability and scalability.

```
mport math
import time

import jax
import jax.numpy as jnp
from jax import jacfwd, jacrev, jit

from microjax.caustics.lightcurve import magnifications as caustic_magnifications
from microjax.point_source import critical_and_caustic_curves

jax.config.update("jax_enable_x64", True)

NUM_POINTS = 2000
NPTS_LIMB = 500

@jit
def get_lightcurve(params: jnp.ndarray):
    """Return times, complex source coordinates, and magnification curve."""
    t0, tE, u0, q, s, alpha, rho, q3, r3, psi = params
    times = t0 + jnp.linspace(-0.5 * tE, tE, NUM_POINTS)
    tau = (times - t0) / tE
    y1 = -u0 * jnp.sin(alpha) + tau * jnp.cos(alpha)
    y2 = u0 * jnp.cos(alpha) + tau * jnp.sin(alpha)
    w_points = (y1 + 1j * y2).astype(jnp.complex128)
    mags = caustic_magnifications(
        w_points,
        rho,
        nlenses=3,
        s=s,
        q=q,
        q3=q3,
        r3=r3,
        psi=psi,
        npts_limb=NPTS_LIMB,
        limb_darkening=False,
    )
    return times, w_points, mags


# JAX evaluates the Jacobian of the magnification curve with respect to the model
# parameters (t0, tE, u0, q, s, alpha, rho, q3, r3, psi).
lightcurve_jacobian = jit(jacfwd(lambda p: get_lightcurve(p)[2]))

DEFAULT_PARAMS = jnp.array(
    [
        0.0,
        10.0,
        0.1,
        0.1,
        1.1,
        jnp.deg2rad(50.0),
        0.02,
        0.03,
        jnp.abs(0.3 + 1.2j),
        jnp.arctan2((0.3 + 1.2j).imag, (0.3 + 1.2j).real),
    ],
    dtype=jnp.float64,
)


def evaluate_lightcurve(params: jnp.ndarray = DEFAULT_PARAMS):
    """Compute the light curve and Jacobian for the provided parameter vector."""
    times, w_points, mags = get_lightcurve(params)
    jac = lightcurve_jacobian(params)
    return times, w_points, mags, jac


if __name__ == "__main__":
    params = DEFAULT_PARAMS
    start = time.time()
    times, w_points, mags, jac = evaluate_lightcurve(params)
    mags.block_until_ready()
    runtime = time.time() - start
    print(
        f"Computed {mags.shape[0]} magnification samples "
        f"and Jacobian in {runtime:.3f} s (JIT compiled)."
    )
    start = time.time()
    _, _ , _ , _ = evaluate_lightcurve(params)
    _.block_until_ready() 
    runtime = time.time() - start
    print(
        f"Computed {mags.shape[0]} magnification samples "
        f"and Jacobian in {runtime:.3f} s (2nd call)."
    )
```
The runtimes are 

> Computed 2000 magnification samples and Jacobian in 61.202 s (JIT compiled).
> Computed 2000 magnification samples and Jacobian in 424.764 s (2nd call).

I am also not sure why the second call takes much time. And, the output plot is here.

<img width="1299" height="1427" alt="Image" src="https://github.com/user-attachments/assets/fb48bd9c-79b5-4310-9eea-cd7d679056e6" />

**Comment by CoastEgo** [comment]
Hi Shota, I also noticed this issue when developing `microlux`. I think the problem is that the convergence of the magnification doesn't promise the convergence of the gradient. Our solution is to design a new error estimator that is related to the gradient error to ensure convergence. See Equation 18 and Figure 5  in [microlux paper](https://iopscience.iop.org/article/10.3847/1538-3881/adb1b2/pdf).

In 'Caustic', they use the image distance (see extended source](https://github.com/fbartolic/caustics/blob/main/src/caustics/extended_source.py#L114)) to guide the adaptive sampling, which is only related to the magnification convergence. I presume that developing a gradient-related error will solve this problem. I hope this will help you.

**Comment by ShotaMiyazaki94** [comment]
Hi @CoastEgo, thanks a lot! that seems really helpful. I will take a close look at that in the microlux paper. I agree that introducing a gradient error estimator is probably the right direction.

By the way, do you plan to extend `microlux` to support more complex lens systems, or to continue maintaining it in the future? I noticed that `caustics` already supports triple-lens modeling, though it seems to need further optimization. `microlux` looks already well-optimized for the binary case, so I was just curious about its future direction, purely out of interest.

**Comment by CoastEgo** [comment]
Hi @ShotaMiyazaki94, I will continue maintaining `microlux`(I hope so), but we currently don't plan to extend this to the triple lens case. 
The implementation of `microlux` relies on the well-established adaptive contour integral package for binary lensing, `VBBinaryLensing`, which is robust and fast in practice. But the extension for the triple lens case, `VBMicrolensing`, still seems to face some problems related to numerical instability, like inversion of the 10-th polynomial. The Jax-based `VBMicrolensing` will face more problems because the contour integral is hard to parallelize. But these two shortcomings seem to be avoided by your code, which is promising for complex lens systems, in my understanding

**Comment by CoastEgo** [comment]
Hi @ShotaMiyazaki94, thanks for your insights as well. I will reach out if there are any new ideas on this. Likewise, I'd be happy to hear any further thoughts from you in the future. Cheers
